src/main/scala/com/microsoft/hyperspace/index/rules/HyperspaceFilter.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.ActiveSparkSession
+import com.microsoft.hyperspace.index.IndexLogEntry
+
+trait HyperspaceFilter {
+
+  /**
+   * @return Failure reason for filtered out indexes.
+   */
+  def reason: String
+}
+
+/**
+ * Filter used in CandidateIndexCollector.
+ */
+trait SourceFilter extends HyperspaceFilter with ActiveSparkSession {
```

Comment:
This sounds like we are filtering the source plan whereas we are filtering indexes. How about `IndexFilterOnSourcePlan` and `IndexFilterOnQueryPlan`? (assuming you are ok with `IndexFilter`)